### PR TITLE
PE-275 - Add custom reports api djangoapp

### PR DESCRIPTION
**Pull request by ericfab179**
This PR adds the djangoapp "custom_reports" developed by Zeus Learning for edx-platform.

### CODE REVIEW

File: common/djangoapps/custom_reports/api/__init__.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+"""
+init.py file for api
+"""
+
+# Flag for enabling Threading in report generation
+ENABLE_THREADING_IN_RG = True
```
Comment (andrey-canon) [inline common/djangoapps/custom_reports/api/__init__.py:6]: should it be a microsite setting ?

Comment (ryangadams) [reply common/djangoapps/custom_reports/api/__init__.py:6]: I assume threading is a server level decision? We'd probably use threading unless the server didn't support it? So probably not a microsite setting? 

I really don't know though.

**Comment by ryangadams** [comment]
Should the custom_reports be a separate git repo that we include into edx-platform via pip (or some other approach)? I'm thinking about our future plans to migrate to Hawthorn/Ironwood/Juniper

File: common/djangoapps/custom_reports/api/__init__.py (hunk 1 above)
Comment (andrey-canon) [reply common/djangoapps/custom_reports/api/__init__.py:6]: I dont know the context for this django app, but this settings should be configurable (django settings, microsite settings, waffle flag.....etc) rather that change the code to modify the behavior.

File: common/djangoapps/custom_reports/api/__init__.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+"""
+init.py file for api
+"""
+
+# Flag for enabling Threading in report generation
+ENABLE_THREADING_IN_RG = True
+
+# Total number of threads to be started at a time for weekly report generation
+TOTAL_THREADS_IN_WEEKLY_REPORT = 5
```
Comment (andrey-canon) [inline common/djangoapps/custom_reports/api/__init__.py:9]: same comment  above

File: common/djangoapps/custom_reports/api/__init__.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+"""
+
+# Flag for enabling Threading in report generation
+ENABLE_THREADING_IN_RG = True
+
+# Total number of threads to be started at a time for weekly report generation
+TOTAL_THREADS_IN_WEEKLY_REPORT = 5
+
+# Reports directory name
+REPORTS_DIR = '/edx/var/edxapp/data/reports'
```
Comment (andrey-canon) [inline common/djangoapps/custom_reports/api/__init__.py:12]: maybe store this on S3

File: common/djangoapps/custom_reports/views.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+from django.http import HttpResponse
+from django.http import JsonResponse
+from django.views.decorators.csrf import csrf_exempt
+from django.views.decorators.http import require_POST
+
+
+# sample callback
+@csrf_exempt
+@require_POST
+def my_callback(request):
```
Comment (andrey-canon) [inline common/djangoapps/custom_reports/views.py:13]: I think this method has to be redefined, what is the purpose of this ?

File: common/djangoapps/custom_reports/views.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    print(request.body)
```
Comment (andrey-canon) [inline common/djangoapps/custom_reports/views.py:14]: ????

File: common/djangoapps/custom_reports/utils.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+import requests
+import uuid, csv, json, math, copy
+from api import REPORTS_DIR
+from django.conf import settings
+from django.core.validators import URLValidator
+import logging
+
+AUDIT_LOG = logging.getLogger("audit")
+
+class CompletionCheckThread(Thread):
```
Comment (andrey-canon) [inline common/djangoapps/custom_reports/utils.py:20]: Why don't use tasks with celery ??

**Comment by ericfab179** [comment]
> Should the custom_reports be a separate git repo that we include into edx-platform via pip (or some other approach)? I'm thinking about our future plans to migrate to Hawthorn/Ironwood/Juniper

@ryangadams Yes maybe another approach is better. @andrey-canon could it be a pulgin ?

**Comment by andrey-canon** [comment]
> > Should the custom_reports be a separate git repo that we include into edx-platform via pip (or some other approach)? I'm thinking about our future plans to migrate to Hawthorn/Ironwood/Juniper
> 
> @ryangadams Yes maybe another approach is better. @andrey-canon could it be a pulgin ?

@ryangadams  @ericfab179 yeah, this should be a plugin in order to make  the migrations easier,  but first we should test if the plugins works right on this platform version and then re-factor the app.

File: common/djangoapps/custom_reports/utils.py (hunk 6 above)
Comment (ryangadams) [reply common/djangoapps/custom_reports/utils.py:20]: I'm guessing they don't know about celery. I don't.

Is this an acceptable approach anyway?

File: common/djangoapps/custom_reports/views.py (hunk 4 above)
Comment (ryangadams) [reply common/djangoapps/custom_reports/views.py:13]: Ah, it's a sample callback to show how the web hook would work. I think we can safely ignore it?

File: common/djangoapps/custom_reports/api/__init__.py (hunk 1 above)
Comment (ryangadams) [reply common/djangoapps/custom_reports/api/__init__.py:6]: understood. I'll make them change that.

**Comment by ryangadams** [comment]
Ok, here's what I'm going to ask Zeus to change and re-submit:

1. rename djangoapp from proversity to proversity_custom_reports (different from what we've got here but better I think)
2. write the report to S3 instead of application server disk
3. extract TOTAL_THREADS_IN_WEEKLY_REPORT and ENABLE_THREADING_IN_RG to settings with 5 and true as default values

I'm not going to request shifting to celery or a plugin, but we might need to consider that approach in future.

**Pull request closed without merging**